src/v/storage/disk_log_impl.cc

```diff
@@ -1117,91 +1118,79 @@ ss::future<> disk_log_impl::truncate(truncate_config cfg) {
-                  model::no_timeout);
-            });
-      })
-      .then(
-        [this, cfg, pidx](std::optional<std::pair<model::offset, size_t>> phs) {
-            if (!phs) {
-                return ss::make_exception_future<>(
-                  std::runtime_error(fmt_with_ctx(
-                    fmt::format,
-                    "User asked to truncate at:{}, with initial physical "
-                    "position of:{}, but internal::offset_to_filepos_consumer "
-                    "could not translate physical offsets. Log state: {}",
-                    cfg,
-                    pidx,
-                    *this)));
-            }
-            auto [prev_last_offset, file_position] = phs.value();
-            auto last = _segs.back();
-            if (file_position == 0) {
-                _segs.pop_back();
-                return remove_segment_permanently(
-                  last, "truncate[post-translation]");
-            }
-            _probe.remove_partition_bytes(last->size_bytes() - file_position);
-            return _readers_cache->evict_truncate(cfg.base_offset)
-              .then([this,
-                     prev_last_offset = prev_last_offset,
-                     file_position = file_position,
-                     last,
-                     phs](readers_cache::range_lock_holder cache_lock) {
-                  return last->truncate(prev_last_offset, file_position)
-                    .handle_exception([last, phs, this](std::exception_ptr e) {
-                        vassert(
-                          false,
-                          "Could not truncate:{} logical max:{}, physical "
-                          "offset:{} on "
-                          "segment:{} - log:{}",
-                          e,
-                          phs->first,
-                          phs->second,
-                          last,
-                          *this);
-                    })
-                    .finally([cache_lock = std::move(cache_lock)] {});
-              });
-        });
+
+    // an unchecked reader is created which does not enforce the logical
+    // starting offset. this is needed because we really do want to read
+    // all the data in the segment to find the correct physical offset.
+    auto reader = co_await make_unchecked_reader(
+      log_reader_config(start, cfg.base_offset, cfg.prio));
+    auto phs = co_await std::move(reader).consume(
+      internal::offset_to_filepos_consumer(cfg.base_offset, initial_size),
+      model::no_timeout);
+
+    if (!phs) {
+        throw std::runtime_error(fmt_with_ctx(
+          fmt::format,
+          "User asked to truncate at:{}, with initial physical "
+          "position of:{}, but internal::offset_to_filepos_consumer "
+          "could not translate physical offsets. Log state: {}",
+          cfg,
+          pidx,
+          *this));
+    }
+    auto [prev_last_offset, file_position] = phs.value();
+    auto last_ptr = _segs.back();
```

Comment:
we check if segs was empty earlier, but we don't again here despite having many scheduling points between the two points in the code. is there a lock or something that ensures that segs is not now empty?

Reply:
will add a check, good point